caffe2/opt/backend_cutting.cc

```diff
@@ -0,0 +1,416 @@
+  for (const auto& op : net_opt.op()) {
+    auto op_node = g->createNode();
+    for (const auto& input : op.input()) {
+      if (!tensor_map.count(input)) {
+        auto tensor = caffe2::make_unique<nom::repr::Tensor>(input);
+        tensor_map[input] =
+            g->createNode(unique_dyn_cast<nom::repr::NeuralNetData>(tensor));
+      }
+
+      auto tensor_node = tensor_map[input];
+      g->createEdge(tensor_node, op_node);
+    }
+
+    for (const auto& output : op.output()) {
+      if (!tensor_map.count(output)) {
+        auto tensor = caffe2::make_unique<nom::repr::Tensor>(output);
+        tensor_map[output] =
+            g->createNode(unique_dyn_cast<nom::repr::NeuralNetData>(tensor));
+      }
+      auto tensor_node = tensor_map[output];
+      g->createEdge(op_node, tensor_node);
+    }
+
+    op_node->resetData(nom::converters::convertOperator(op));
+    auto op_ref = dyn_cast<nom::repr::NeuralNetOperator>(op_node->data().get());
+    CAFFE_ENFORCE(op_node->data());
+    auto device_name = op.device_option().node_name();
+    if (device_name != "") {
+      auto device = caffe2::make_unique<nom::repr::DeviceAnnotation>(device_name);
+      op_ref->setAnnotation(std::move(device));
+    } else {
+      op_ref->setAnnotation(caffe2::make_unique<nom::repr::Annotation>());
+    }
+    op_ref->getMutableAnnotation()->setSaved((void *)&op);
+  }
+}
+
+void PruneUnrefereredNodes(nom::repr::NNGraph* g) {
+  std::vector<NodeRef> to_delete;
+  for (auto node : g->getMutableNodes()) {
+    if (!nom::repr::nn::hasProducer(node) &&
+        !nom::repr::nn::hasConsumer(node)) {
+      to_delete.push_back(node);
+    }
+  }
+  for (auto i : to_delete) {
+    g->deleteNode(i);
+  }
+}
+
+} // namespace
+
+caffe2::NetDef OptimizeForBackend(
+    const caffe2::NetDef& net,
+    std::function<bool(const caffe2::OperatorDef&)> supports,
+    std::function<caffe2::NetDef(const caffe2::NetDef&)> transform_func) {
+  auto nn = nom::converters::convertFromCaffe2Proto(net);
+  auto& dfg = nn.dataFlow;
+
+  // Initialize the group info and figure out the external/input output
+  VisitorContext context(supports);
+  std::vector<NodeRef> external_inputs;
+  std::vector<NodeRef> external_outputs;
+  for (auto node : dfg.getMutableNodes()) {
+    auto* node_data = node->data().get();
+    context.infos.emplace(
+        std::piecewise_construct,
+        std::forward_as_tuple(node),
+        std::forward_as_tuple(node->getInEdges().size(), -1));
+
+    if (!isa<nom::repr::NeuralNetOperator>(node_data)) {
+      if (!nom::repr::nn::hasProducer(node)) {
+        external_inputs.push_back(node);
+      }
+      if (!nom::repr::nn::hasConsumer(node)) {
+        external_outputs.push_back(node);
+      }
+    }
+  }
+
+  // Find unsupported and supported groups of nodes alernatively
+  context.frontier.clear();
+  context.current_group.clear();
+  context.find_supported = false;
+  std::vector<TransformSubgraph> subs;
+  for (std::vector<NodeRef> frontier(
```

Comment:
perhaps this would be a bit more readable as do{} while?

Reply:
Probably. I like the `GO` style where we only have `for`. :D